src/Compilers/CSharp/Test/Semantic/Semantics/LambdaDiscardParametersTests.cs

```diff
@@ -0,0 +1,477 @@
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    static void M()
+    {
+        int _ = 0;
+        System.Func<string, string, int> f = (_, _) => _++;
+        System.Func<long, string, long> f2 = (_, a) => _++;
+    }
+}");
+            // Note that naming one of the parameters seems irrelevant but results in a binding change
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);
+            var underscores = tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Where(p => p.ToString() == "_").ToArray();
+            Assert.Equal(2, underscores.Length);
+
+            var localSymbol = model.GetSymbolInfo(underscores[0]).Symbol;
+            Assert.Equal("System.Int32 _", localSymbol.ToTestDisplayString());
+            Assert.Equal(SymbolKind.Local, localSymbol.Kind);
+
+            var parameterSymbol = model.GetSymbolInfo(underscores[1]).Symbol;
+            Assert.Equal("System.Int64 _", parameterSymbol.ToTestDisplayString());
+            Assert.Equal(SymbolKind.Parameter, parameterSymbol.Kind);
+        }
+
+        [Fact]
+        public void DiscardParameters_NotInScope_BindToOutsideLocal_Nested()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    static void M()
+    {
+        int _ = 0;
+        System.Func<string, string, int> f = (_, _) =>
+        {
+            System.Func<string, string, int> f2 = (_, _) => _++;
+            return f2(null, null);
+        };
+    }
+}");
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);
+            var underscore = tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Where(p => p.ToString() == "_").Single();
+
+            var localSymbol = model.GetSymbolInfo(underscore).Symbol;
+            Assert.Equal("System.Int32 _", localSymbol.ToTestDisplayString());
+            Assert.Equal(SymbolKind.Local, localSymbol.Kind);
+        }
+
+        [Fact]
+        public void DiscardParameters_NotInScope_DeclareLocalNamedUnderscoreInside()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    static void M()
+    {
+        System.Func<string, string, long> f = (_, _) => { long _ = 0; return _++; };
+        System.Func<string, string, long> f2 = (_, a) => { long _ = 0; return _++; };
+    }
+}");
+            // Note that naming one of the parameters seems irrelevant but results in a binding change
+            comp.VerifyDiagnostics(
+                // (7,65): error CS0136: A local or parameter named '_' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
+                //         System.Func<string, string, long> f2 = (_, a) => { long _ = 0; return _++; };
```

Comment:
>System.Func<string, string, long> f2 = (_, a) => { long _ = 0; return _++; }; [](start = 27, length = 77)

Putting all declarations on the same line makes it hard to determine about which declaration we complain.